pkg/authentication/tokens.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2024 Canonical Ltd.
+// SPDX-License-Identifier: AGPL-3.0
+
+package authentication
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/coreos/go-oidc/v3/oidc"
+	"go.opentelemetry.io/otel/trace"
+	"golang.org/x/oauth2"
+
+	"github.com/canonical/identity-platform-admin-ui/internal/logging"
+)
+
+type JWKSTokenVerifier struct {
+	verifier providerVerifier
+	tracer   trace.Tracer
+	logger   logging.LoggerInterface
+}
+
+func verifyJWT(ctx context.Context, rawJwt string, verifier providerVerifier) (*Principal, error) {
+	i, err := verifier.Verify(ctx, rawJwt)
+	if err != nil {
+		return nil, err
+	}
+
+	return NewPrincipalFromClaims(i)
+}
+
+func (j *JWKSTokenVerifier) VerifyAccessToken(ctx context.Context, rawAccessToken string) (*Principal, error) {
+	_, span := j.tracer.Start(ctx, "authentication.JWKSTokenVerifier.VerifyAccessToken")
+	defer span.End()
+
+	return verifyJWT(ctx, rawAccessToken, j.verifier)
+}
+
+func (j *JWKSTokenVerifier) VerifyIDToken(ctx context.Context, rawIDToken string) (*Principal, error) {
+	_, span := j.tracer.Start(ctx, "authentication.JWKSTokenVerifier.VerifyIDToken")
+	defer span.End()
+
+	return verifyJWT(ctx, rawIDToken, j.verifier)
+}
+
+func NewJWKSTokenVerifier(provider ProviderInterface, clientID string, tracer trace.Tracer, logger logging.LoggerInterface) *JWKSTokenVerifier {
+	j := new(JWKSTokenVerifier)
+	j.tracer = tracer
+	j.logger = logger
+
+	j.verifier = provider.Verifier(&oidc.Config{ClientID: clientID})
+
+	return j
+}
+
+type UserinfoTokenVerifier struct {
+	clientID string
+	provider ProviderInterface
+	verifier providerVerifier
+
+	tracer trace.Tracer
+	logger logging.LoggerInterface
+}
+
+type claims struct {
+	Audience []string `json:"aud"`
+}
+
+func (u *UserinfoTokenVerifier) VerifyAccessToken(ctx context.Context, rawAccessToken string) (*Principal, error) {
+	_, span := u.tracer.Start(ctx, "authentication.UserinfoTokenVerifier.VerifyAccessToken")
+	defer span.End()
+
+	info, err := u.provider.UserInfo(ctx, oauth2.StaticTokenSource(&oauth2.Token{AccessToken: rawAccessToken}))
```

Comment:
same here for otelhttp, we will need to add something to the context to make sure it s using a traceable http client

Reply:
Done for every `context` use in the OAuth2 implementation. The httpclient with Otel transport is shared.